Added "auto select best class" feature for qx.data.marshal.Json

schnepe2 opened the pull request:
Attempt to fix issue #9193

schnepe2 committed:
Added "auto select best class" feature

level420 commented:
@schnepe2 Thank you for the PR.

It seems that one of your tests fail constantly on Firefox and Chrome:
```
Chrome 53.0.2785 (Linux 0.0.0) qx.test.data.store Json:testOwnClassWith FAILED
	Not the given class used!: Expected 'qx.test.AB' but found 'qx.data.model.a~b'!
		qx.core.AssertionError.constructor:43:38
		qx.core.Assert.__fail:71:19
		qx.core.Assert.assertEquals:177:33
		qx.type.BaseString.assertEquals:87:22
		anonymous()
		qx.dev.unit.TestResult.run:237:36
		qx.dev.unit.TestCase.resume:100:35
		anonymous()
		qx.event.dispatch.Direct.dispatchEvent:134:34
		wrappedFunction()
		qx.event.Manager.prototype.dispatchEvent()
		qx.event.Registration.fireNonBubblingEvent()
		qx.core.Object.prototype.fireDataEvent()
		qx.data.store.Json.prototype._onSuccess()
		...
```

schnepe2 commented:
That "old" test assumes a name for the class that might not be correct.
Strange that this didn't happen on my machine. These tests are head-less, right? I'll see what I have to do to make them pass again.

schnepe2 committed:
Adjusted store test

schnepe2 commented:
The test that failed **now** (qx.test.ui.basic Label:testApplyWebFont) is nothing I have done (I think)

schnepe2 committed:
"automatic" vs. "forced" mode feature.

level420 commented:
@schnepe2 Yes the failing test ```testApplyWebFont``` is failing because of timing issues with travis. See https://github.com/qooxdoo/qooxdoo/issues/9202

But there is another failure here:
```
Running lint...
  - Warning: qx.data.marshal.Json (118, 11): Declared but unused variable or parameter: 'clazz'
Lint warnings are treated as errors! Please fix!
```

schnepe2 committed:
Fixed lint warning

schnepe2 commented:
About all the tests that failed **now**: I have no idea what happens there.
And unfortunately I have spend more time on this than my boss would allow me to do, so either someone from the core team looks over this or we have to wait until I can steal some more time.

schnepe2 commented:
F***k ! I hate the complexity of git! It distracts me from the most obvious things!
I'll try a fix of the fix...

schnepe2 committed:
Bugfix (regression from last commit)

schnepe2 committed:
Fixed lint warning

level420 commented:
The PR is passing the tests. Start voting/approval.

johnspackman commented on framework/source/class/qx/data/marshal/Json.js:
```diff
@@ -301,7 +317,15 @@ qx.Class.define("qx.data.marshal.Json",
+        if (!clazz)
+        {
+          // Extra check for possible bubble-event feature inconsistency
+          var className_ = className.replace(/~/g, '"');
```
Please can you use a different name for the variable - the underscore does not help describe it's purpose, eg `bubbleClassName` ?

johnspackman commented on framework/source/class/qx/data/marshal/Json.js:
```diff
@@ -114,14 +114,12 @@ qx.Class.define("qx.data.marshal.Json",
-      var hash = this.__jsonToHash(data);  // without bubble event feature
-      var clazz = "qx.data.model." + hash; //    "      "      "      "
-
-      var hash_ = hash.replace(/"/g, '~');
```
also please rename the underscores here

schnepe2 committed:
variable rename

schnepe2 commented:
@johnspackman Done (not tested of cause!).
Although I think those kind of qooxdoo-internal naming conventions could also have been easily fixed in trunk by you :wink:

oetiker commented:
@schnepe2 we do not fix things in master ... changes get into master via PR ... and sure, someone might have done a cleanup PR after merging yours, but then again ... since we have no cleanup crew on staff ...

level420 commented:
@schnepe2 If I (qooxdoo member) want to change something in master, I have to go through the same procedure you've gone. Create a PR ... Wait for travis runs to be happy ... Start voting ... Wait for comments or change requests ... modify ... sync the PR ...  

The only thing I can do what you can't is merging. But NOT my own PRs.

schnepe2 committed:
- Fixed issue with very simple model (only one property, so no glue character distinguishing possible)
- Now a heat ("♥") post-fix indicates a class with bubbling event.
- Variable name fixed (bubbleClassName is actually noBubbleClassName)

schnepe2 commented:
The corner case of a very simple model (having only one property (like `{a:"a"}`) ) wasn't working
(no glue character to distinguish what kind of class we have)
I therefore changed the "decoration" of the hash by adding a heart prefix ("♥") to model-classes which support the bubble event.

oetiker commented:
@schnepe2 have you seen how many fixes I made to my #9194 PR until it was merged? We do this for two reasons:

a) to ensure that qx master is always clean (to the best of our knowledge), so whenever you pull master you can be sure you get 'the good stuff'

b) often we have to merge the patches for master also with the current stable branch (aka 5.0.x) if every change is one PR, then this is much easier.

level420 commented:
4 approvals, more than 2 business days
merging
Thank you @schnepe2
